### PR TITLE
[ogr] Insure that the connections pool always carry on open options

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR fixes #45534 by insuring open options within the data source URI are always passed onto the OGR connections pool. A test was added to cover this scenario.

Power to CSV datasets! :wink:

### CODE REVIEW

**Comment by nirvn** [comment]
(unrelated test failure)

**Comment by nirvn** [comment]
@rouault , I'd love your review on this. IMHO, this should make it into 3.22.0. While it does touch a crucial part of QGIS (the core OGR provider), I think the change is safe, and actually makes that OGR provider options feature work as it should on feature iterator for the first time since its introduction a few versions back :)

File: src/core/providers/ogr/qgsogrproviderutils.cpp [hunk 1]
```diff
@@ -957,7 +957,17 @@ QString QgsOgrProviderUtils::connectionPoolId( const QString &dataSourceURI, boo
...
+    {
+      // Preserve open options so pooled connections always carry those on
+      QString openOptions;
+      static thread_local QRegularExpression openOptionsRegex( QStringLiteral( "((?:\\|option:(?:[^|]*))+)" ) );
+      QRegularExpressionMatch match = openOptionsRegex.match( dataSourceURI );
+      if ( match.hasMatch() )
+      {
+        openOptions = match.captured( 1 );
+      }
+      return filePath + openOptions;
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrproviderutils.cpp:969]: it should be safe to pass open options for the non-file case too (return dataSourceURI + openOptions)

File: src/core/providers/ogr/qgsogrproviderutils.cpp [hunk 2]
```diff
@@ -957,7 +957,17 @@ QString QgsOgrProviderUtils::connectionPoolId( const QString &dataSourceURI, boo
     QString filePath = dataSourceURI.left( dataSourceURI.indexOf( QLatin1Char( '|' ) ) );
     QFileInfo fi( filePath );
     if ( fi.isFile() )
-      return filePath;
+    {
+      // Preserve open options so pooled connections always carry those on
+      QString openOptions;
+      static thread_local QRegularExpression openOptionsRegex( QStringLiteral( "((?:\\|option:(?:[^|]*))+)" ) );
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrproviderutils.cpp:963]: couldn't we use QgsOgrProviderMetadata::decodeUri() instead ?

Comment (nirvn) [reply src/core/providers/ogr/qgsogrproviderutils.cpp:963]: @rouault , we could, but I was worried it'd be a bit too costly for no real reason (we're not re-encoding afterwards, we're not in need of other bits either).

File: src/core/providers/ogr/qgsogrproviderutils.cpp (hunk 1 above)
Comment (rouault) [reply src/core/providers/ogr/qgsogrproviderutils.cpp:969]: oh, right.

**Review by rouault — APPROVED** [review]

**Pull request merged**